src/gdextension/render.h

```diff
@@ -0,0 +1,226 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <Magnum/Trade/ImageData.h>
+#include <Magnum/Trade/MeshData.h>
+#include <godot_cpp/variant/rid.hpp>
+#include <longeron/id_management/registry_stl.hpp>
+#include <osp/drawing/drawing_fn.h>
+
+namespace osp::draw
+{
+
+enum class TexGdId : uint32_t
+{
+};
+enum class MeshGdId : uint32_t
+{
+};
+
+using TexGdStorage_t  = Storage_t<TexGdId, godot::RID>;
+using MeshGdStorage_t = Storage_t<MeshGdId, godot::RID>;
+
+/**
+ * @brief Main renderer state and essential GL resources
+ *
+ * This may be shared between scenes
+ */
+struct RenderGd
+{
+
+    // Fullscreen Triangle
+    MeshGdId m_fullscreenTri;
+    // FullscreenTriShader                 m_fullscreenTriShader{Corrade::NoCreate};
+
+    // Offscreen Framebuffer
+    TexGdId m_fboColor;
+    // Magnum::GL::Renderbuffer            m_fboDepthStencil{Corrade::NoCreate};
+    // Magnum::GL::Framebuffer             m_fbo{Corrade::NoCreate};
+
+    // Renderer-space GL Textures
+    lgrn::IdRegistryStl<TexGdId> m_texIds;
+    TexGdStorage_t               m_texGd;
+
+    // Renderer-space GL Meshes
+    lgrn::IdRegistryStl<MeshGdId> m_meshIds;
+    MeshGdStorage_t               m_meshGd;
+
+    // Associate GL Texture Ids with resources
+    IdMap_t<ResId, TexGdId>        m_resToTex;
+    IdMap_t<TexGdId, ResIdOwner_t> m_texToRes;
+
+    // Associate GL Mesh Ids with resources
+    IdMap_t<ResId, MeshGdId>        m_resToMesh;
+    IdMap_t<MeshGdId, ResIdOwner_t> m_meshToRes;
+
+    // godot resources
+    godot::RID scenario;
+    godot::RID viewport;
+};
+
+struct ACompTexGd
+{
+    TexId   m_scnId{ lgrn::id_null<TexId>() };
+    TexGdId m_gdId{ lgrn::id_null<TexGdId>() };
+};
+
+struct ACompMeshGd
+{
+    MeshId   m_scnId{ lgrn::id_null<MeshId>() };
+    MeshGdId m_glId{ lgrn::id_null<MeshGdId>() };
+};
+
+using MeshGdEntStorage_t     = KeyedVec<DrawEnt, ACompMeshGd>;
+using TexGdEntStorage_t      = KeyedVec<DrawEnt, ACompTexGd>;
+using InstanceGdEntStorage_t = KeyedVec<DrawEnt, godot::RID>;
+
+/**
+ * @brief OpenGL specific rendering components for rendering a scene
+ */
+struct ACtxSceneRenderGd
+{
+    MeshGdEntStorage_t     m_meshId;
+    TexGdEntStorage_t      m_diffuseTexId;
+    InstanceGdEntStorage_t m_instanceId;
+    godot::RID             m_scenario;
```

Comment:
scene?

"scenario" more refers to a loadable scene

Reply:
It's mostly because the object behind the RID is a scenario for Godot. Which is not the same as a scene (that also exists in Godot). It has nothing to do with the OSP scenario, but I guess it should be more clear somehow.